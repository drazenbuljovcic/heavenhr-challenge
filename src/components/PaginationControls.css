:local(.pagination) {
  display: flex;
  justify-content: center;
}

:local(.invisiblePaginationItem) {
  opacity: 0;
  pointer-events: none;
}

:local(.invisibleShrunkenPaginationItem) {
  opacity: 0;
  pointer-events: none;
  width: 0;
  height: 0;
}

:local(.activePage) {
  color: green;
  font-size: 1.65em;
  font-weight: 700;
  padding: 0 12px;
}

:local(.paginationLegend) {
  color: white;
  text-align: center;
  font-size: 1.2em;
}