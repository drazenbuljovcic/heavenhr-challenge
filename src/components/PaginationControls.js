import React from 'react';
import PropTypes from 'prop-types';
import { times } from 'lodash';

export default class PaginationControls extends React.Component {
  //#region MARK: Super Methods
  constructor(props, context) {
    super(props, context);

    this.handlePaginationToNextPage = this.handlePaginationToNextPage.bind(this);
    this.handlePaginationToPreviousPage = this.handlePaginationToPreviousPage.bind(this);
    this.handlePaginationToPageNumber = this.handlePaginationToPageNumber.bind(this);
  }
  //#endregion

  //# MARK: Pagination navigation handlers
  handlePaginationToNextPage() {
    const { currentPage, changePaginationPage } = this.props;

    changePaginationPage(currentPage + 1);
  }

  handlePaginationToPreviousPage() {
    const { currentPage, changePaginationPage } = this.props;

    changePaginationPage(currentPage- 1);
  }

  handlePaginationToPageNumber(pageNumber) {
    this.props.changePaginationPage(pageNumber);
  }
  //#endregion

  render() {
    return (
      <div>
        <ul className="pagination">
          <li>
            <button href="#" className="btn btn-default" aria-label="Previous" onClick={this.handlePaginationToPreviousPage}>
              <span aria-hidden="true">&laquo;</span>
            </button>
          </li>
          {
            times(this.props.totalPages, (i) => (
              <li key={i}>
                <button href="#" className="btn btn-default" onClick={this.handlePaginationToPageNumber.bind(null, i)}>
                  {i + 1}
                </button>
              </li>
            ))
          }
          <li>
            <button href="#" className="btn btn-default" aria-label="Next" onClick={this.handlePaginationToNextPage}>
              <span aria-hidden="true">&raquo;</span>
            </button>
          </li>
        </ul>
      </div>
    );
  }
}

PaginationControls.propTypes = {
  totalPages: PropTypes.number.isRequired,
  currentPage: PropTypes.number.isRequired,
  changePaginationPage: PropTypes.func.isRequired,
}