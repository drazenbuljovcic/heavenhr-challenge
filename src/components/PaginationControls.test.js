import React from 'react';

import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-15';

import PaginationControls from './PaginationControls';

configure({ adapter: new Adapter() });

describe('Pagination Controls', () => {
  const defaultProps = {
    totalPages: 0,
    currentPage: 0,
    totalCount: 0,
    startingItemIndex: 0,
    endingItemIndex: 0,
    handlePaginationToFirstPage: jest.fn(),
    handlePaginationToLastPage: jest.fn(),
    handlePaginationToNextPage: jest.fn(),
    handlePaginationToPreviousPage: jest.fn(),
    handlePaginationToPageNumber: jest.fn(),
  };
  const component = shallow(<PaginationControls {...defaultProps}/>);

  it('should match it\'s empty snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should navigate to first page when backward button is pressed', () => {
    component.find('button[aria-label="Backward"]').simulate('click');
    expect(defaultProps.handlePaginationToFirstPage).toHaveBeenCalled();
  });

  it('should navigate to previous page when previous button is pressed', () => {
    component.find('button[aria-label="Previous"]').simulate('click');
    expect(defaultProps.handlePaginationToPreviousPage).toHaveBeenCalled();
  });

  it('should navigate to page number when page number button is pressed', () => {
    const tmpComponent = shallow(<PaginationControls {...defaultProps} totalPages={2}/>);
    tmpComponent.find('button[aria-label="Page 2"]').simulate('click');
    expect(defaultProps.handlePaginationToPageNumber).toHaveBeenCalledWith(1);
  });

  it('should navigate to next page when next button is pressed', () => {
    component.find('button[aria-label="Next"]').simulate('click');
    expect(defaultProps.handlePaginationToNextPage).toHaveBeenCalled();
  });

  it('should navigate to last page when forward button is pressed', () => {
    component.find('button[aria-label="Forward"]').simulate('click');
    expect(defaultProps.handlePaginationToLastPage).toHaveBeenCalled();
  });

  it('should not display previous and first page buttons if pagination is on first page', () => {
    const tmpComponent = shallow(<PaginationControls {...defaultProps} totalPages={2} currentPage={0} />);
    expect(toJson(tmpComponent)).toMatchSnapshot();
  });

  it('should not display next and forward page buttons if pagination is on last page', () => {
    const tmpComponent = shallow(<PaginationControls {...defaultProps} totalPages={2} currentPage={1} />);
    expect(toJson(tmpComponent)).toMatchSnapshot();
  });

  it('should display legend on the bottom', () => {
    const tmpComponent = shallow(<PaginationControls {...defaultProps} totalCount={5} startingItemIndex={2} endingItemIndex={3} />);
    expect(toJson(tmpComponent)).toMatchSnapshot();
  })
});