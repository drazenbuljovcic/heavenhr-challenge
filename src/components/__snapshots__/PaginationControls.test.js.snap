exports[`Pagination Controls should display legend on the bottom 1`] = `
<div>
  <ul
    className="undefined pagination">
    <li
      aria-hidden={true}
      className="undefined">
      <button
        aria-label="Backward"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-fast-backward" />
      </button>
    </li>
    <li
      aria-hidden={true}
      className="undefined">
      <button
        aria-label="Previous"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-step-backward" />
      </button>
    </li>
    <li
      aria-hidden={false}
      className="">
      <button
        aria-label="Next"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-step-forward" />
      </button>
    </li>
    <li
      aria-hidden={false}
      className="">
      <button
        aria-label="Forward"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-fast-forward" />
      </button>
    </li>
  </ul>
  <p>
    Showing 
    <strong>
      2
       - 
      3
    </strong>
     out of 
    <strong>
      5
    </strong>
  </p>
</div>
`;

exports[`Pagination Controls should match it's empty snapshot 1`] = `
<div>
  <ul
    className="undefined pagination">
    <li
      aria-hidden={true}
      className="undefined">
      <button
        aria-label="Backward"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-fast-backward" />
      </button>
    </li>
    <li
      aria-hidden={true}
      className="undefined">
      <button
        aria-label="Previous"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-step-backward" />
      </button>
    </li>
    <li
      aria-hidden={false}
      className="">
      <button
        aria-label="Next"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-step-forward" />
      </button>
    </li>
    <li
      aria-hidden={false}
      className="">
      <button
        aria-label="Forward"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-fast-forward" />
      </button>
    </li>
  </ul>
  <p>
    Showing 
    <strong>
      0
       - 
      0
    </strong>
     out of 
    <strong>
      0
    </strong>
  </p>
</div>
`;

exports[`Pagination Controls should not display next and forward page buttons if pagination is on last page 1`] = `
<div>
  <ul
    className="undefined pagination">
    <li
      aria-hidden={false}
      className="">
      <button
        aria-label="Backward"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-fast-backward" />
      </button>
    </li>
    <li
      aria-hidden={false}
      className="">
      <button
        aria-label="Previous"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-step-backward" />
      </button>
    </li>
    <li
      className=""
      key="0">
      <button
        aria-label="Page 1"
        className="btn btn-default"
        onClick={[Function]}>
        1
      </button>
    </li>
    <li
      className=""
      key="1">
      <button
        aria-label="Page 2"
        className="btn btn-default undefined"
        onClick={[Function]}>
        2
      </button>
    </li>
    <li
      aria-hidden={true}
      className="undefined">
      <button
        aria-label="Next"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-step-forward" />
      </button>
    </li>
    <li
      aria-hidden={true}
      className="undefined">
      <button
        aria-label="Forward"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-fast-forward" />
      </button>
    </li>
  </ul>
  <p>
    Showing 
    <strong>
      0
       - 
      0
    </strong>
     out of 
    <strong>
      0
    </strong>
  </p>
</div>
`;

exports[`Pagination Controls should not display previous and first page buttons if pagination is on first page 1`] = `
<div>
  <ul
    className="undefined pagination">
    <li
      aria-hidden={true}
      className="undefined">
      <button
        aria-label="Backward"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-fast-backward" />
      </button>
    </li>
    <li
      aria-hidden={true}
      className="undefined">
      <button
        aria-label="Previous"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-step-backward" />
      </button>
    </li>
    <li
      className=""
      key="0">
      <button
        aria-label="Page 1"
        className="btn btn-default undefined"
        onClick={[Function]}>
        1
      </button>
    </li>
    <li
      className=""
      key="1">
      <button
        aria-label="Page 2"
        className="btn btn-default"
        onClick={[Function]}>
        2
      </button>
    </li>
    <li
      aria-hidden={false}
      className="">
      <button
        aria-label="Next"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-step-forward" />
      </button>
    </li>
    <li
      aria-hidden={false}
      className="">
      <button
        aria-label="Forward"
        className="btn btn-default"
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-fast-forward" />
      </button>
    </li>
  </ul>
  <p>
    Showing 
    <strong>
      0
       - 
      0
    </strong>
     out of 
    <strong>
      0
    </strong>
  </p>
</div>
`;
